### PR TITLE
e2e test: verify that user can type right after inserting the quote block

### PR DESCRIPTION
**Pull request by oandregal**
## What?

Introduces a new test for making sure an user can type after inserting a quote.

## Why?

The original test was migrated to Playwright at https://github.com/WordPress/gutenberg/pull/40216 In doing so, we changed some of the original behavior. This PR brings it back by adding a new test.

## How?

Before the migration to Playwright, in adding a quote, we did so in two steps:

- added the block
- typed some content

We did so because it reflects what an user would do, and also because these steps cover the following case: we wanted to make sure that users can type right after the block is inserted.

This was lost during the migration to Playwright, where we added the block with the contents directly.

## Testing Instructions

Make sure e2e tests pass.

### CODE REVIEW

**Comment by kevin940726** [comment]
> but shouldn't we respect the existing steps when migrating tests?

It depends on the test. If the test is about inserting blocks and ensuring users could type right after inserting them, then yes, absolutely. However, if the test is about switching to the plain style of the quote block, which is what this test is about, then I think we're allowed to use some shortcuts.

The same applies to the purpose of those REST helpers to clear/set states between tests. Ideally, we would want to do what end users do in our tests, but often that's too slow and sometimes flaky. To prevent testing the same thing over and over again, we create some shortcuts/mocks to do them in a much faster way. It's covered in the [best practices](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/README.md#restify-actions-to-clear-or-set-states) of the documentation.

Unless I misunderstood what this test is about, I think we should test the abovementioned behavior in another test, specifically for this kind of interaction. WDYT?

_(Additionally, the new `insertBlock` behavior doesn't insert blocks from the inserter anymore, but rather inserts them from calling `wp.data.dispatch('core/block-editor').insertBlock()` in the console. Hence, the original behavior will still be lost even if we add back `page.type` after the insertion.)_

**Comment by oandregal** [comment]
Thanks for the explanation. Though I'm now a bit more confused :sweat_smile: Sorry if I'm missing context, I don't have tons of experience with the way e2e tests are set up. I guess my core question is: doesn't using different code paths than a user would defeats the purpose of the e2e tests? How e2e tests built this way are different than unit tests?

**Comment by oandregal** [comment]
> Unless I misunderstood what this test is about, I think we should test the abovementioned behavior in another test, specifically for this kind of interaction. WDYT?

I'm happy to add more specific tests, but for the kind of flow I wanted to test (focus, typing, etc, what we had before) I'd need to make sure that we're actually using the user code paths (the inserter, etc). Have looked around but don't find a way to do that, is there any? I guess I'd need to write a new utility that actually uses the inserter?

**Comment by youknowriad** [comment]
I agree with @kevin940726 on principle but I don't think a migration should not change how tests work unless we also cover the part that was changed. 

> (Additionally, the new insertBlock behavior doesn't insert blocks from the inserter anymore, but rather inserts them from calling wp.data.dispatch('core/block-editor').insertBlock() in the console. Hence, the original behavior will still be lost even if we add back page.type after the insertion.)

This for me is very problematic. `insertBlock` is used in so many tests and in so many different ways that if we decided to change its behavior, it's going to have an enormous impact on the coverage. Every time we migrate a test that uses it, we should be thinking about what we're losing in terms of coverage (and this PR is the perfect example).

I'd have preferred personally to avoid changing any behavior for the migration itself, and having dedicated PRs to change the behavior because that way we can think clearly about the impact and don't miss anything while migrating.

For this PR, it's probably better to just add a new test covering the typing but I'm really concerned about the "change everything approach" we seem to be taking with the migration.

File: test/e2e/specs/editor/various/quote-type-after-inserting.spec.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+/**
```
Comment (youknowriad) [inline test/e2e/specs/editor/various/quote-type-after-inserting.spec.js]: Should this test be just "editor/blocks/quote" (if could include the puppeteer ones we already have later)

Comment (oandregal) [reply test/e2e/specs/editor/various/quote-type-after-inserting.spec.js]: Makes a lot more sense :+1:

Comment (oandregal) [reply test/e2e/specs/editor/various/quote-type-after-inserting.spec.js]: https://github.com/WordPress/gutenberg/pull/40466/commits/620abcf1ed64f25f3d72cd7a4da97b08b0a24323

**Review by youknowriad — APPROVED** [review]

**Comment by kevin940726** [comment]
> doesn't using different code paths than a user would defeats the purpose of the e2e tests? How e2e tests built this way are different than unit tests?

It depends on what you want to test. I'm not good with words, so probably the [documentation](https://docs.cypress.io/guides/references/best-practices#Organizing-Tests-Logging-In-Controlling-State) of Cypress and the referenced video there could explain it better 😅 .

> I don't think a migration should not change how tests work unless we also cover the part that was changed.

Ideally, I totally agree with this. But in practice, this will probably slow down the migration process by a lot, and we'll not be using some of the benefits that Playwright brings until very late in the process.

> Every time we migrate a test that uses it, we should be thinking about what we're losing in terms of coverage (and this PR is the perfect example).

IMO, code path coverage isn't what's important in e2e tests; rather, the coverage of what the test is about is more important. The migration process is never going to cover all existing coverage, there are always going to be some differences, and that's okay.

> I'd have preferred personally to avoid changing any behavior for the migration itself, and having dedicated PRs to change the behavior because that way we can think clearly about the impact and don't miss anything while migrating.

This is an option, but the fact that so many tests depend on `e2e-test-utils` makes it complicated. The package depends on Puppeteer, and even though they are very much alike, there are still many fundamental differences that make them incompatible in many ways. We could migrate all utils from `e2e-test-utils` to their one-to-one mapping in `e2e-test-utils-playwright` as closely as possible. But the approach I took is to gradually rewrite part of the tests during the migration to follow the new [best practices](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/README.md). The plan is to migrate tests to strictly follow their original _testing purposes_ as much as possible, and safely rewrite those intermediate steps (`activateTheme`, `insertBlock`, etc) that are not relevant to the outcome of the tests.

IMO, we should focus on what we want to test, or else it's difficult for contributors to change any of the existing tests, and thus making the purpose of the tests blur and not readable to others. (Everything is important will make nothing important.)

As you suggested before, [deleting tests is a best practice](https://riad.blog/2020/07/21/deleting-tests-is-a-best-practice/), we shouldn't be afraid of removing exiting coverage if that's not even the part of the test itself.

**Comment by youknowriad** [comment]
> As you suggested before, [deleting tests is a best practice](https://riad.blog/2020/07/21/deleting-tests-is-a-best-practice/), we shouldn't be afraid of removing exiting coverage if that's not even the part of the test itself.

Good one :)

**Pull request merged**